web/packages/teleport/src/Discover/SelectResource/SelectResource.tsx

```diff
@@ -346,58 +346,73 @@ function sortResourcesByKind(
 
 const aBeforeB = -1;
 const aAfterB = 1;
+const aEqualsB = 0;
+
+/**
+ * Evaluates the predicate and prioritizes the element matching the predicate over the element that
+ * doesn't.
+ *
+ * @example
+ * comparePredicate({color: 'green'}, {color: 'red'}, (el) => el.color === 'green') // => -1 (a before b)
+ * comparePredicate({color: 'red'}, {color: 'green'}, (el) => el.color === 'green') // => 1  (a after  b)
+ * comparePredicate({color: 'blue'}, {color: 'pink'}, (el) => el.color === 'green') // => 0  (both are equal)
+ */
+function comparePredicate<ElementType>(
+  a: ElementType,
+  b: ElementType,
+  predicate: (resource: ElementType) => boolean
+): -1 | 0 | 1 {
+  const aMatches = predicate(a);
+  const bMatches = predicate(b);
+
+  if (aMatches && !bMatches) {
+    return aBeforeB;
+  }
+
+  if (bMatches && !aMatches) {
+    return aAfterB;
+  }
+
+  return aEqualsB;
+}
 
 export function sortResources(
   resources: ResourceSpec[],
   preferences: UserPreferences
 ) {
   const { preferredResources, hasPreferredResources } =
     getPrioritizedResources(preferences);
+  const platform = getPlatform();
 
   const sortedResources = [...resources];
   const accessible = sortedResources.filter(r => r.hasAccess);
   const restricted = sortedResources.filter(r => !r.hasAccess);
 
   // Sort accessible resources by 1. os 2. preferred 3. guided and 4. alphabetically
   accessible.sort((a, b) => {
-    let aPreferred,
-      bPreferred = false;
-    if (hasPreferredResources) {
-      aPreferred = preferredResources.includes(
-        resourceKindToPreferredResource(a.kind)
-      );
-      bPreferred = preferredResources.includes(
-        resourceKindToPreferredResource(b.kind)
-      );
-    }
-
-    const platform = getPlatform();
+    const compareAB = (predicate: (r: ResourceSpec) => boolean) =>
+      comparePredicate(a, b, predicate);
 
     // Display platform resources first
-    if (a.platform === platform && b.platform !== platform) {
-      return aBeforeB;
-    }
-    if (a.platform !== platform && b.platform === platform) {
-      return aAfterB;
+    const prioritizeUserPlatform = compareAB(r => r.platform === platform);
+    if (prioritizeUserPlatform) {
+      return prioritizeUserPlatform;
```

Comment:
Let me know if this refactor is helpful in any way. It might not seem like much when viewed from the point of view of this single commit.

However, when I started adding more checks for sortingConnect My Computer, I pulled the plug and refactored the existing conditionals to use the `comparePredicate` function instead.

It does make the code slightly less elegant – we have to check for `0` and continue sorting, we have to use this custom method, we have to use `compareAB` instead of `compare(a, b, …)` because Prettier would format that in a rather verbose way.

But the biggest advantage is that we no longer have to write the conditional for `a` and then repeat its inverse for `b`. I feel like that part of the code was most vulnerable to subtle bugs.